### PR TITLE
RM-7346: SharedSource migration

### PR DESCRIPTION
**Pull request by dkudernatsch**
https://re-motion.atlassian.net/browse/RM-7346

### CODE REVIEW

File: SharedSource/Core/Utilities/ArgumentUtility.cs [hunk 1]
```diff
@@ -19,7 +19,7 @@
 using System.Diagnostics;
 using System.Reflection;
 using JetBrains.Annotations;
-
+#nullable disable
```
Comment (iam-mholle) [inline SharedSource/Core/Utilities/ArgumentUtility.cs]: At the current state, `ArgumentUtility` does not provide the needed flow enhancements.

```csharp
void Do (object a) {}

void A (object? a)
{
  ArgumentUtility.CheckNotNull ("a", a);

  Do (a); // [CS8604] Possible null reference argument for parameter 'a' in 'void A(object a)'.
}
```

I guess our main use case is providing runtime argument checks for **non-nullable** parameters, but the flow should still be enhanced.

File: SharedSource/Core/Utilities/Assertion.cs [hunk 2]
```diff
@@ -16,8 +16,9 @@
 // 
 using System;
 using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
 using JetBrains.Annotations;
-
+#nullable enable
 // ReSharper disable once CheckNamespace
 namespace Remotion.Utilities
```
Comment (iam-mholle) [inline SharedSource/Core/Utilities/Assertion.cs]: `IsTrue` and `IsFalse` need `DoesNotReturnIfAttribute` for the assertion condition.

e.g.:
```csharp
[AssertionMethod]
public static void IsTrue ([DoesNotReturnIf (false), AssertionCondition (AssertionConditionType.IS_TRUE)] bool assertion, string message)
{
  IsTrue (assertion, message, s_emptyArguments);
}
```

File: SharedSource/Core/Utilities/ArgumentUtility.cs (hunk 1 above)
Comment (iam-mholle) [reply SharedSource/Core/Utilities/ArgumentUtility.cs]: Unforturnately I can't comment on unchanged lines, so I'll post everything here.

```csharp
[AssertionMethod]
[return: NotNullIfNotNull ("actualValue")]
public static T CheckNotNull<T> (
    [InvokerParameterName] string argumentName,
    [System.Diagnostics.CodeAnalysis.NotNull, AssertionCondition (AssertionConditionType.IS_NOT_NULL), NoEnumeration] T actualValue)
{
  // ReSharper disable CompareNonConstrainedGenericWithNull
  if (actualValue == null)
      // ReSharper restore CompareNonConstrainedGenericWithNull
    throw new ArgumentNullException (argumentName);

  return actualValue;
}
```

`[return: NotNullIfNotNull ("actualValue")]` is needed on every method doing null checks and returning the value if not null.

`[System.Diagnostics.CodeAnalysis.NotNull]` fixes the wrong flow addressed in the first comment. This needs to be added on every parameter that is null checked.

Comment (iam-mholle) [inline SharedSource/Core/Utilities/ArgumentUtility.cs]: This should be `#nullable enable`. Please change and fix errors to allow a second review.

File: SharedSource/Development/RhinoMocks/UnitTesting/ListArgExtensions.cs [hunk 3]
```diff
@@ -34,22 +36,24 @@ static partial class ListArgExtensions
     /// <summary>
     /// Similiar to <see cref="ListArg{T}.Equal"/> but without considering the order of the elements in the collection.
     /// </summary>
+    [return: MaybeNull]
```
[the remark this answers is not in the document]
Comment (dkudernatsch) [reply SharedSource/Development/RhinoMocks/UnitTesting/ListArgExtensions.cs]: I think so, because otherwise the call site is not aware that this can return null.

File: SharedSource/Core/FunctionalProgramming/Maybe.cs [hunk 4]
```diff
@@ -18,7 +18,7 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
-
+#nullable enable
 // ReSharper disable once CheckNamespace
 namespace Remotion.FunctionalProgramming
```
[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/FunctionalProgramming/Maybe.cs]: MayBe will die. All re-motion internal usages are just syntactical sugar for null propagation, null annotation, and null coalesing. Public usages, if there are, have to be reviewed for their breaking change impact. Dropping Maybe for internal usages could be done right now. Possibly even before we do the null rewrite?

File: SharedSource/Core/Utilities/Assertion.cs [hunk 5]
```diff
@@ -46,110 +47,110 @@ static partial class Assertion
...
     [AssertionMethod]
-    public static void DebugIsNotNull<T> ([AssertionCondition (AssertionConditionType.IS_NOT_NULL)] T obj, string message)
+    public static void DebugIsNotNull<T> ([AssertionCondition (AssertionConditionType.IS_NOT_NULL), MaybeNull] T obj, string message)
     {
       IsNotNull (obj, message, s_emptyArguments);
     }
 
     [AssertionMethod]
-    public static T IsNotNull<T> ([AssertionCondition (AssertionConditionType.IS_NOT_NULL)] T obj, string message)
+    public static T IsNotNull<T> ([AssertionCondition (AssertionConditionType.IS_NOT_NULL), MaybeNull] T obj, string message)
```
[the remark this answers is not in the document]
Comment (dkudernatsch) [reply SharedSource/Core/Utilities/Assertion.cs]: I think all of them should be `[AllowNull, NotNull]` so it is ok to pass nullable input, but after the the method returns they are NotNull.
(at least i think that's what this is supposed to mean)

[the remark this answers is not in the document]
Comment (iam-mholle) [reply SharedSource/Core/Utilities/Assertion.cs]: As `T` is unconstrained, `AllowNull` is redundant in my opinion.

File: SharedSource/Development/RhinoMocks/UnitTesting/ListArgExtensions.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (iam-mholle) [reply SharedSource/Development/RhinoMocks/UnitTesting/ListArgExtensions.cs]: Right

File: SharedSource/Core/Utilities/MemberInfoEqualityComparer.cs [hunk 6]
```diff
@@ -47,7 +48,7 @@ sealed partial class MemberInfoEqualityComparer<T> : IEqualityComparer<T> where
     /// The idea for this method, but not the code, was taken from http://blogs.msdn.com/b/kingces/archive/2005/08/17/452774.aspx.
     /// </para>
     /// </returns>
-    public bool Equals (T one, T two)
+    public bool Equals ([AllowNull] T one, [AllowNull] T two)
```
[the remark this answers is not in the document]
Comment (dkudernatsch) [reply SharedSource/Core/Utilities/MemberInfoEqualityComparer.cs]: that doesnt work because of the constraint on `T: MemberInfo`, as `T?` would violate nullability of MemberInfo.
As `T: MemberInfo` implies `T: class` however `bool Equals(T? one, T? two)` is possible

File: SharedSource/Core/FunctionalProgramming/Maybe.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (iam-mholle) [reply SharedSource/Core/FunctionalProgramming/Maybe.cs]: Added [RM-7396 - Remove usages of Maybe](https://re-motion.atlassian.net/browse/RM-7396).

File: SharedSource/Core/Utilities/MemberInfoEqualityComparer.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (iam-mholle) [reply SharedSource/Core/Utilities/MemberInfoEqualityComparer.cs]: You're right, must have overlooked this. `bool Equals(T? one, T? two)` it is.

File: SharedSource/Core/Utilities/ReferenceEqualityComparer.cs [hunk 7]
```diff
@@ -28,15 +29,15 @@ namespace Remotion.Utilities
   /// and not just equal for reference checking purposes.      
   /// </summary>      
   /// <typeparam name="T">the type of object to check</typeparam>  
-  partial class ReferenceEqualityComparer<T> : IEqualityComparer<T>
+  partial class ReferenceEqualityComparer<T> : IEqualityComparer<T> where T: notnull
```
Comment (iam-mholle) [inline SharedSource/Core/Utilities/ReferenceEqualityComparer.cs]: My bad. I think a `class` constraint and `public bool Equals (T? x, T? y)` would be better. what do you think?

File: SharedSource/Development/RhinoMocks/UnitTesting/Threading/LockingDecoratorTestHelper.cs [hunk 8]
```diff
@@ -46,19 +47,21 @@ public LockingDecoratorTestHelper (T lockingDecorator, object lockObject, T inne
...
+        where TResult : notnull
     {
       ArgumentUtility.CheckNotNull ("action", action);
       ArgumentUtility.CheckNotNull ("fakeResult", fakeResult);
 
       ExpectSynchronizedDelegation (action, fakeResult, r => Assert.That (r, Is.EqualTo (fakeResult)));
     }
 
-    public void ExpectSynchronizedDelegation<TResult> (Func<T, TResult> action, TResult fakeResult, Action<TResult> resultChecker)
+    public void ExpectSynchronizedDelegation<TResult> (Func<T, TResult> action, TResult fakeResult, Action<TResult> resultChecker) where TResult: notnull
```
Comment (iam-mholle) [inline SharedSource/Development/RhinoMocks/UnitTesting/Threading/LockingDecoratorTestHelper.cs]: Formatting.

```suggestion
    public void ExpectSynchronizedDelegation<TResult> (Func<T, TResult> action, TResult fakeResult, Action<TResult> resultChecker) where TResult : notnull
```

Multiple occurrences.

File: SharedSource/Core/RoslynAnnotations/NullableAnnotations.cs [hunk 9]
```diff
@@ -1,5 +1,4 @@
-﻿﻿
-#define INTERNAL_NULLABLE_ATTRIBUTES
+﻿#define INTERNAL_NULLABLE_ATTRIBUTES
 // https://github.com/dotnet/corefx/blob/48363ac826ccf66fbe31a5dcb1dc2aab9a7dd768/src/Common/src/CoreLib/System/Diagnostics/CodeAnalysis/NullableAttributes.cs
```
Comment (MichaelKetting) [inline SharedSource/Core/RoslynAnnotations/NullableAnnotations.cs:2]: need a matching LM file and "correct" / complete license header. compare with ReSharper annotations.

File: SharedSource/Core/Utilities/Assertion.cs [hunk 10]
```diff
@@ -160,23 +164,23 @@ public static T IsNotNull<T> ([AssertionCondition (AssertionConditionType.IS_NOT
     }
 
     [AssertionMethod]
-    public static void IsNull ([AssertionCondition (AssertionConditionType.IS_NULL)] object obj, string message)
+    public static void IsNull ([AssertionCondition (AssertionConditionType.IS_NULL)] object? obj, string message)
```
[the remark this answers is not in the document]
Comment (iam-mholle) [reply SharedSource/Core/Utilities/Assertion.cs:167]: Something like `[DoesNotReturnWhenNull]` would really be helpful here.

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs [hunk 11]
```diff
@@ -396,7 +396,7 @@ public static IEnumerable<T> Concat<T> (this IEnumerable<T> source, T item)
       using (var enumerator = source.GetEnumerator())
       {
         if (!enumerator.MoveNext())
-          return default (TSource);
+          return default (TSource)!;
```
[the remark this answers is not in the document]
Comment (iam-mholle) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:399]: I guess the method should have `[return: MaybeNull]`.

File: SharedSource/Core/Utilities/Assertion.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/Utilities/Assertion.cs:167]: ah, wait, you can do this with ContractAnnotation at least: null => throw (exact sytnax to be checked with docs)

[the remark this answers is not in the document]
Comment (iam-mholle) [reply SharedSource/Core/Utilities/Assertion.cs:167]: `"null => halt"`, but that does not help with roslyn right?

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:399]: okay, because TSource can already be nullable type.

File: SharedSource/Core/Utilities/Assertion.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/Utilities/Assertion.cs:167]: Nope, but with resharper

[the remark this answers is not in the document]
Comment (dkudernatsch) [reply SharedSource/Core/Utilities/Assertion.cs:167]: i think for roslyn we can add a `[NotNull]` postcondition to obj, as it signals that after the method returns the parameter can no longer be null
https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-05-15.md#simple-postconditions

[the remark this answers is not in the document]
Comment (iam-mholle) [reply SharedSource/Core/Utilities/Assertion.cs:167]: It should be `null` here tho.

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/Utilities/Assertion.cs:167]: Am fuzzy right now. IsNull throws if obj is not null. Thus, notnull will halt, correct?

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs [hunk 12]
```diff
@@ -139,7 +139,7 @@ static partial class EnumerableExtensions
       ArgumentUtility.CheckNotNull ("predicate", predicate);
       ArgumentUtility.CheckNotNull ("createNoMatchingElementException", createNoMatchingElementException);
 
-      TSource result = default (TSource);
+      TSource result = default (TSource)!;
```
[the remark this answers is not in the document]
Comment (dkudernatsch) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:142]: unfortunately this is currently not supported by NRT, there are plans to allow
either a T: unconstrained or T?? to signify the default value of an unconstrained type param

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs [hunk 13]
```diff
@@ -101,7 +101,7 @@ static partial class EnumerableExtensions
       ArgumentUtility.CheckNotNull ("source", source);
       ArgumentUtility.CheckNotNull ("createEmptySequenceException", createEmptySequenceException);
 
-      TSource result = default (TSource);
+      TSource result = default (TSource)!;
```
[the remark this answers is not in the document]
Comment (dkudernatsch) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:104]: Single always returns the same type as the list is constructed with e.g a list with nullable entries has string? as TSource

File: SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs [hunk 14]
```diff
@@ -152,7 +153,7 @@ public void CreateSequence_WhileNotNull ()
       var third = new Element (3, second);
       var fourth = new Element (4, third);
 
-      IEnumerable<Element> actual = fourth.CreateSequence (e => e.Parent);
+      IEnumerable<Element> actual = fourth.CreateSequence (e => e.Parent!);
```
[the remark this answers is not in the document]
Comment (dkudernatsch) [reply SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs:156]: This works only for `CreateSequence<TSource> (this TSource source, Func<TSource, TSource?> nextElementSelector)` as it has a class constraint, however the called `CreateSequence<TSource> (this TSource source, Func<TSource, TSource> nextElementSelector, Func<TSource, bool> predicate)` is not constrained

[the remark this answers is not in the document]
Comment (dkudernatsch) [reply SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs:156]: also i think it is currently not supported to annotate type parameters using the attributes in any form

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:143]: this means there is a compiler error if we do not set the ! at the end? Even though TSource can be nullable?

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:143]: We need to add a test that shows that the sequence can contain null values and everything still compiles. Plus tests to show that Single also works with a sequence of valuetype and nullable value type

File: SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs]: Element.Parent is declared as nullable, having to wrongly specify null-forgiving here is at least weird. this value can be null, saying it cannot be null is incorrect.

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs]: should we discuss those in the next devjourfixe so everyone can have fun?
@iam-mholle

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (iam-mholle) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:143]: > this means there is a compiler error if we do not set the ! at the end? Even though TSource can be nullable?

`default(TSource)` is `null` for a reference type, nullable or not. `TSource` could be a non-nullable reference type, which would get assigned `null`, that's why it does not compile without `!`.

[the remark this answers is not in the document]
Comment (iam-mholle) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:143]: > We need to add a test that shows that the sequence can contain null values and everything still compiles. Plus tests to show that Single also works with a sequence of valuetype and nullable value type

As `TSource` is unconstrained there is no difference to the code before NRT, so I do not see a value in adding tests (they would just test the compiler, right?), as this is no runtime feature. Am I missing something?

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:143]: They test that the syntax works for all the specified cases. Given there is some weird magic involved, we should make sure that all supported scenarios compile and read syntactically clean on the caller side.

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:143]: > 
> 
> > this means there is a compiler error if we do not set the ! at the end? Even though TSource can be nullable?
> 
> `default(TSource)` is `null` for a reference type, nullable or not. `TSource` could be a non-nullable reference type, which would get assigned `null`, that's why it does not compile without `!`.

So it doesn't affect the typing of TSource, it's just an "ignore this and move on"? I was always under the impression it's the other way around, saying that a possibly null-value won't be null and thus it's okay to assign to a not-nullable variable/parameter/etc.

[the remark this answers is not in the document]
Comment (dkudernatsch) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:143]: The actual type of `default(TSource)` cant be represented in current C#, because for value types it would be `TSource` and for ref types `TSource?`. The null forgiving operator converts the expression to type `TSource` regardless of actual nullability. 
There is currently a draft for the syntax ``TSource?? result = default(TSource);` for unconstrained type parameters.

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:143]: Yes. 

I just checked the code base and we've got about 15 usages of generic default values. For now, my suggestion stands if it wouldn't be cleaner to just disable NRT for the assignment line using a #enable / #disable instead of using the null-forgiving operator. It just feels cleaner/easier to understand to openly express that NRT doesn't work in this generic situation. null-forgiving is okay in test code when we need to initialize fields that we want to declare as not-nullable, but here in the production code, this particualr semantics would be turned on it's head: the null-forgiving operator would be used to assign a potentially null value as "not null" to a variable of undetermined null-ability. the casual reader would then see both types of usage and step into this exact problem I fell into.

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs [hunk 15]
```diff
@@ -249,7 +249,7 @@ public static IEnumerable<TSource> CreateSequence<TSource> (this TSource source,
         this TSource source,
         Func<TSource, TSource> nextElementSelector,
         Func<TSource, bool> predicate,
-        [CanBeNull] IEqualityComparer<TSource> equalityComparer,
+        [CanBeNull] IEqualityComparer<TSource>? equalityComparer,
```
[the remark this answers is not in the document]
Comment (iam-mholle) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:253]: We decided to leave them in, but I can't remember the reasoning - I guess it was the need of nullability info for the projects, due to a resharper issue/missing feature. If the need of adapting **ImplicitNullability** arises, wouldn't it be beneficial to add the attributes as a post-compilation step, matching the generated roslyn nullability attributes?

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:253]: @iam-mholle 
not all projects will use C# 8 (or a c# 8-compiler-based Version of Visual Studio)
Wasn't there also the problem that R# itself ignored NRTs in base-projects if the own project isn't NRT enabled?

So, we need R# annotations. How to get them? Option one, leave them in. Problem is, for the 'big' code base, that's going to grow old, fast because we'd also have to maintain them forward. Option two, we update the ImplictNullablity addin. I do not want to do a post-compile (also that's an external decision, too). So, the R# addin would be a much simpler solution.

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (dkudernatsch) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:143]: Did we decide to use `!` everywhere to be consistent with corefx etc. ?

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:143]: Yes.

File: SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs [hunk 16]
```diff
@@ -182,14 +212,14 @@ public void CreateSequence_WhilePredicateEvaluatesTrue ()
       var third = new Element (3, second);
       var fourth = new Element (4, third);
 
-      IEnumerable<Element> actual = fourth.CreateSequence (e => e.Parent, e => e != first);
+      IEnumerable<Element?> actual = ((Element?)fourth).CreateSequence (e => e!.Parent, e => e != first);
```
Comment (MichaelKetting) [inline SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs:215]: Those calls / tests need another go-around, sorry. There should not be a need to do these acrobatics. That's quite breaking and annoying on the user side

File: SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs [hunk 17]
```diff
@@ -182,14 +212,14 @@ public void CreateSequence_WhilePredicateEvaluatesTrue ()
...
-      IEnumerable<Element> actual = fourth.CreateSequence (e => e.Parent, e => e != first);
+      IEnumerable<Element?> actual = ((Element?)fourth).CreateSequence (e => e!.Parent, e => e != first);
       Assert.That (actual.ToArray (), Is.EqualTo (new[] { fourth, third, second }));
     }
 
     [Test]
     public void CreateSequence_WhilePredicateEvaluatesTrue_WithNull ()
     {
-      IEnumerable<Element> actual = ((Element)null).CreateSequence (e => e.Parent, e => e != null);
+      IEnumerable<Element> actual = ((Element?)null).CreateSequence (e => e!.Parent, e => e != null)!;
```
Comment (MichaelKetting) [inline SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs:222]: Can't we write this as 
((Element)null!).CreateSequence  (e => e.Parent, e= e!= null)

File: SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs [hunk 18]
```diff
@@ -198,7 +228,7 @@ public void CreateSequence_WhilePredicateEvaluatesTrue_WithSingleElement ()
     {
       var element = new Element (0, null);
 
-      IEnumerable<Element> actual = element.CreateSequence (e => e.Parent, e => e != null);
+      IEnumerable<Element> actual = element.CreateSequence (e => e!.Parent!, e => e != null);
```
Comment (MichaelKetting) [inline SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs:231]: There should be no NFO here needed. That's public, this needs to be clean.

File: SharedSource/UnitTests/Utilities/MemberInfoFromExpressionUtilityTest.cs [hunk 19]
```diff
@@ -210,7 +212,7 @@ public void GetMethod_Static ()
     [Test]
     public void GetMethod_Static_VoidGeneric ()
     {
-      var member = MemberInfoFromExpressionUtility.GetMethod (() => DomainType.StaticVoidGenericMethod<SampleType> (null));
+      var member = MemberInfoFromExpressionUtility.GetMethod (() => DomainType.StaticVoidGenericMethod<SampleType?> (null));
```
Comment (MichaelKetting) [inline SharedSource/UnitTests/Utilities/MemberInfoFromExpressionUtilityTest.cs:215]: Hmm...it doesn't really make a differnce for the test, but I do wonder if it would be more expressive to write null! and not make the type parameters nullable. that's here and with the other file that uses the DomainType helper class. Null is just used because we don't have a value but we do mean an actual instance and want the not-nullable version.

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs [hunk 20]
```diff
@@ -190,12 +191,12 @@ public static IEnumerable<TSource> CreateSequence<TSource> (this TSource source,
...
     /// until the <paramref name="nextElementSelector"/> returns <see langword="null" />.
     /// </returns>
-    public static IEnumerable<TSource> CreateSequence<TSource> (this TSource source, Func<TSource, TSource> nextElementSelector)
+    public static IEnumerable<TSource> CreateSequence<TSource> (this TSource source, Func<TSource, TSource?> nextElementSelector)
         where TSource : class
     {
       ArgumentUtility.CheckNotNull ("nextElementSelector", nextElementSelector);
 
-      return CreateSequence (source, nextElementSelector, e => e != null);
+      return CreateSequence (source, nextElementSelector!, e => e != null)!;
```
[the remark this answers is not in the document]
Comment (dkudernatsch) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:199]: Otherwise `Func<T, T?>` wouldn't match Func<T, T> in the overload. This is necessary because we can't specify the nullability of T in Func<T, T> when T is unbounded.

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:199]: Nope, still don't understand. nextElementSelector is a delegate. How does the nullablity of the Delegate's type parameters figure into the nullablity of the delegate instance itself? No, didn't check it out to try, so I don't know the error message etc.

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:199]: Okay, and second question: why do we have a NFO on the return value of the nested CreateSequence call? That shouldn't be nullable anyway, so why the NFO?

File: SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs (hunk 18 above)
Comment (dkudernatsch) [reply SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs:231]: If we supply any other predicate this code would produce a NullRefException. Thats why we have an explicit overload that only works with class and can verify that those null values wont make to the nextElementSelector. 
Its definitely not ideal, but i think its reasonable here that you have to tell the compiler that you check for null manually.
(the first ! is not needed and can be removed)

File: SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs (hunk 17 above)
Comment (dkudernatsch) [reply SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs:222]: CreateSequence would then require `Func<T, T>` but `e => e.Parent` is `Func<T, T?>` and would again require a NFO

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (dkudernatsch) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:199]: the `!` operator basically erases Nullability information of a type so `Func<T, T?>` gets converted to `Func<T, T>`.
The nextElementSelector here has to return null sometime otherwise the sequence won't be terminated. The `!` here is so that we can tell the compiler that we are smarter and check for the null values (in the predicate) that would otherwise throw NREs.

File: SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs (hunk 16 above)
Comment (dkudernatsch) [reply SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs:215]: but this code would actually throw if the sequence ends before e == first

File: SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs (hunk 18 above)
Comment (MichaelKetting) [reply SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs:231]: ah, I .c. This null-check-predicate option is so stupid anyway. Getting a headache...

Of course, now I'm surprised that we can skip the FIRST NFO, instead of the second. The first one's not null only thanks to the predicate, but that we don't know on a compiler level. the second one can actually be null. I'm starting to wonder if this method can actually be annotated for nulalblity beyond the fact that the sequence will never be null.

File: SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (MichaelKetting) [reply SharedSource/Core/FunctionalProgramming/EnumerableExtensions.cs:199]: What the? By supplying the NFO on a variable, we change the way the generic type parameters that are part of the type itself behave, isntead of just how the type as a whole behaves? This is completely...I mean, how? Why? Okay, I'm pretty sure that as a good reason for doing this, but why can't I do some sort of casting or something? THis is just not intuitive :/

File: SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs (hunk 16 above)
Comment (MichaelKetting) [reply SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs:215]: possibly. still having an issue with the element in the sequence becoming a null option when there's none involved. Dev Jour fixe mob session today for CreateSequence?

File: SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs (hunk 18 above)
Comment (dkudernatsch) [reply SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs:231]: `e` is not nullable because we specify (implicitly) that the sequence is of not nullable elements, and e is always an element of the sequence

File: SharedSource/Core/Utilities/EqualityUtility.cs [hunk 21]
```diff
@@ -207,7 +207,7 @@ public static bool EqualsEquatable<T> (T a, T b)
     public static bool EqualsEquatable<T> (T a, object b)
```
Comment (dkudernatsch) [inline SharedSource/Core/Utilities/EqualityUtility.cs:207]: Potential null bug: if parameter `a` is null and `other` is not this throws an NPE, while the overload above works fine.
```csharp
      EqualityUtility.EqualsEquatable ((string) null!,  "");
      EqualityUtility.EqualsEquatable ((string) null!,  (object) "");
```
Line 2 throws a NPE while Line 1 just returns false

Candidate for bugfix ticket? @MichaelKetting

Comment (MichaelKetting) [reply SharedSource/Core/Utilities/EqualityUtility.cs:207]: yes, please. Not that it's particularily relevant. the EU is intended for implementing Equals and GetHashcode for an object, so the first parameter is never null.

File: SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs [hunk 22]
```diff
@@ -126,6 +135,27 @@ public void Single_ThrowCustomException_Empty ()
...
+    {
+      int? actual = _enumerableWithOneNullableValueTypeValue.Single (() => new ApplicationException ("ExpectedText"));
+      Assert.That (actual, Is.Null);
+    }
+    
+    [Test]
+    public void Single_WithOneValueType ()
+    {
+      int? actual = _enumerableWithOneValueTypeValue.Single (() => new ApplicationException ("ExpectedText"));
+      Assert.That (actual, Is.EqualTo(1));
```
Comment (MichaelKetting) [inline SharedSource/UnitTests/FunctionalProgramming/EnumerableExtensionsTest.cs:156]: whitespace.
